### PR TITLE
Add ability to `init` a TS agent

### PR DESCRIPTION
**Pull request by NickHeiner**
I anticipate tweaking the templated file contents once I implement actual agent functionality.

### CODE REVIEW

File: fixieai/agents-ts/src/template.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const agent = new CodeShotAgent(BASE_PROMPT, FEW_SHOTS);
+
+agent.registerFunc((query: Message): string => {
+    if (query.text === "input") {
+        return "output";
+    }
+    else {
+        throw new Error("Invalid input");
+    }
+});
```
Comment (juberti) [inline fixieai/agents-ts/src/template.ts:35]: add newline

File: fixieai/cli/agent/commands.py [hunk 2]
```diff
@@ -138,7 +206,14 @@ def _update_agent_requirements(
     type=str,
     help="Additional requirements for requirements.txt. Can be specified multiple times.",
 )
-def init_agent(handle, description, entry_point, more_info_url, requirement):
+@click.option(
+    "--language",
+    # Set the possible options to be "python" or "TS"
+    type=click.Choice(["python", "TS"], case_sensitive=False),
```
Comment (juberti) [inline fixieai/cli/agent/commands.py:212]: maybe accept ["python", "py", "typescript", "ts]  here

File: fixieai/cli/agent/commands.py [hunk 3]
```diff
@@ -149,61 +224,21 @@ def init_agent(handle, description, entry_point, more_info_url, requirement):
     current_config.more_info_url = more_info_url
     agent_config.save_config(current_config)
 
+    templator = PythonTemplator() if language == "python" else TypeScriptTemplator()
```
Comment (juberti) [inline fixieai/cli/agent/commands.py:227]: probably should fail if an unknown lang is specified

File: fixieai/cli/agent/commands.py [hunk 4]
```diff
@@ -107,8 +104,79 @@ def _update_agent_requirements(
...
+            else:
+                write_requirements = False
 
-@agent.command("init", help="Creates an agent.yaml file.")
+        if write_requirements:
+            with open(REQUIREMENTS_TXT, "wt") as requirements_txt:
+                requirements_txt.writelines(r + "\n" for r in resolved_requirements)
+    
+
+@agent.command("init", help="Creates an agent.yaml fadsfile.")
```
Comment (juberti) [inline fixieai/cli/agent/commands.py:179]:
```suggestion
@agent.command("init", help="Creates an agent.yaml file.")
```

**Review by juberti — APPROVED** [review]
LG with nits

File: fixieai/cli/agent/commands.py (hunk 3 above)
Comment (NickHeiner) [reply fixieai/cli/agent/commands.py:227]: `click` handles that for us:

```
λ poetry run fixie init --language tsadfs
Usage: fixie init [OPTIONS]
Try 'fixie init --help' for help.

Error: Invalid value for '--language': 'tsadfs' is not one of 'python', 'py', 'typescript', 'TS'.
```

File: fixieai/cli/agent/commands.py (hunk 2 above)
Comment (NickHeiner) [reply fixieai/cli/agent/commands.py]: Good idea!

File: README.md [hunk 5]
```diff
@@ -69,10 +69,17 @@ print(response)
...
 response = session.query("What is 38 * 20302?")
 print(response)
 ```
+
+## Contributing
+If you make changes to the Python CLI, use this command to see them locally:
+
+```
+poetry run fixie [...additional args]
+```
```
Comment (NickHeiner) [inline README.md:85]: Maybe this is obvious to Python devs, but it wasn't obvious to me. :smile: This note will make things more accessible for non-Python people.

File: fixieai/agents-ts/package.json [hunk 6]
```diff
@@ -24,7 +24,7 @@
     "fixie"
   ],
   "author": "Fixie Inc.",
-  "license": "ISC",
+  "license": "Apache-2.0",
```
Comment (NickHeiner) [inline fixieai/agents-ts/package.json:27]: Fixing this to match the overall project license.

File: fixieai/cli/agent/commands.py [hunk 7]
```diff
@@ -108,6 +107,117 @@ def _update_agent_requirements(
...
+
+
+class PythonTemplator(AgentTemplator):
+    def get_agent_template_url(self) -> str:
+        return "https://raw.githubusercontent.com/fixie-ai/fixie-examples/main/agents/template.py"
+
+    def get_main_file_extension(self) -> str:
+        return ".py"
+
+    def write_helper_files(self, params: Dict[str, str]) -> None:
```
Comment (NickHeiner) [inline fixieai/cli/agent/commands.py:174]: This is copied, unchanged, from below.

File: fixieai/cli/agent/commands.py [hunk 8]
```diff
@@ -138,71 +274,51 @@ def _update_agent_requirements(
...
-        current_config = agent_config.AgentConfig()
+        current_config = agent_config.AgentConfig(language=language)
     current_config.handle = handle
     current_config.description = description
     current_config.entry_point = entry_point
+    current_config.language = language
     current_config.more_info_url = more_info_url
+
+    # Having this hook here is maybe a little gross.
+    templator.modify_new_agent_config(current_config)
```
Comment (petersalas) [inline fixieai/cli/agent/commands.py:302]: To handle `entry_point` I'd recommend:
- making the type in `AgentConfig` an `Optional[str]` and default value `None`
- removing `prompt` as you've done
- doing something along the lines of
```python
if entry_point:
   current_config.entry_point = entry_point
elif language == "python":
   current_config.entry_point = DEFAULT_PYTHON_ENTRY_POINT
elif language == "typescript":
   current_config.entry_point = DEFAULT_TYPESCRIPT_ENTRY_POINT
```

I think that probably avoids needing to pass the config through the templator

File: fixieai/cli/agent/commands.py [hunk 9]
```diff
@@ -138,71 +274,51 @@ def _update_agent_requirements(
...
-        with open(REQUIREMENTS_TXT, "rt") as requirements_txt:
-            existing_requirements = list(
-                r.strip() for r in requirements_txt.readlines()
-            )
-    except FileNotFoundError:
-        existing_requirements = []
-
-    resolved_requirements = _update_agent_requirements(
-        existing_requirements, list(requirement)
+    templator.write_helper_files(
```
Comment (petersalas) [inline fixieai/cli/agent/commands.py]: IMO it's probably simpler to structure this as:

```python
def init_agent(...):
  # setup the config
  agent_config.foo = bar #etc.
  
  if language == "python":
     init_python_agent(config)
  elif language == "typescript":
     init_typescript_agent(config)
```

Even though there'll probably some similar logging code or whatever between the two functions I don't think they share enough to make the templator abstraction worth it. WDYT?

**Review by petersalas — APPROVED** [review]
LGTM!

**Pull request merged**